### ISSUE TITLE
consider parameterizing `appliedTogetherMessage`s to add application context

### ISSUE THREAD

**Issue by pq** [dart-lang/sdk#44993]
See discussion on https://dart-review.googlesource.com/c/sdk/+/184621.

The rub is some fixes are scoped by context and it would be good to reflect that context in a more specific message when bulk-applying.

For example, the `MAKE_FINAL` message `'Make final where possible in file'` could be improved by specifying that  it is applying to fields which is  the implicit context.

**Comment by bwilkerson** [comment]
The other option, for that specific case, is to split the `FixKind` into two kinds with appropriate messages.

**Comment by pq** [comment]
Multiple fix kinds would be simpler and might communicate better. Are there any downsides to a proliferation?  (I imagine maybe a dozen or so cases.)

**Comment by bwilkerson** [comment]
Yes. The downside is that LSP has support that uses the `id` of the `FixKind`. That support will be impacted. I'm not certain what the actual impact on users it.

@DanTup

**Comment by pq** [comment]
Excellent, thanks.  @DanTup : let me know if you need more context.  I'm happy to do whatever I can to lessen potential impact on you. 👍

**Comment by DanTup** [comment]
Looking at the code, I think there are two places the fix ID is used:

https://github.com/dart-lang/sdk/blob/235292605ecde3d3060d5fb5f265d47a352090b6/pkg/analysis_server/lib/src/lsp/handlers/handler_code_actions.dart#L370-L371

Here it's used to know whether a particular fix can also be applied by the bulk processor, which is what triggers insertion of a "fix all" version of that fix. Changing the IDs shouldn't impact this (as long as it remains the same for the bulk processor and the original fix).

https://github.com/dart-lang/sdk/blob/235292605ecde3d3060d5fb5f265d47a352090b6/pkg/analysis_server/lib/src/lsp/handlers/handler_code_actions.dart#L169

Here, I think `fix.change.id` might also be the fix kind. It's used to allow users to create keybinds for particular types of fixes. I wouldn't expect anyone to keybind a specific fix (usually you would bind groups of them) so the impact of them changing there is probably very low. Additionally, if the change was to add a suffix (eg. something like changing `fix.makeFinal` to `fix.makeFinal.field`) then the keybind would actually still work exactly the same.

I'm happy to look at the changes and/or test them through LSP if you CC me on any related changes.

**Comment by bwilkerson** [comment]
> sdk/pkg/analysis_server/lib/src/lsp/handlers/handler_code_actions.dart
> Here it's used to know whether a particular fix can also be applied by the bulk processor ...

No, that won't be impacted by any changes to the ids.

However, for context, the reason we're looking at this is that we're in the process of implementing a "fix all" mechanism that's hopefully a bit more general (can be used outside of LSP). In thinking about this a bit more a couple of thoughts have emerged.

First, I believe that there are three kinds of fixes (at least for our current purposes):
- Fixes that can safely be applied to a single diagnostic at a single location because the user explicitly chooses to use that fix for that diagnostic. All fixes fall into this set.
- Fixes that can safely be applied to a single diagnostic across an entire file (or larger body of code) because the user explicitly chooses to do so and the fixes won't conflict with each other. This is a subset of the first set. There might be other reasons to exclude fixes from this set that we haven't yet identified.
- Fixes that can safely be applied when the user doesn't have a choice about which fix to apply. This is a subset of the second set, and applies to the `dart fix` bulk-fix style tool.

The `FixProcessor` is focused on the first set and the `BulkFixProcessor` is focused on the third set, so we believe that we need another similar class that can be focused on the second set.

Second, there's an interesting UX question that I think we need to resolve before we go much farther. At any single location we have both a specific diagnostic code and a specific kind of fix. But sometimes the kind of fix that can be applied to a specific diagnostic code can change based on the context in which the diagnostic was found. And, sometimes the same fix kind can be used to fix multiple diagnostic codes. So, when we "fix all" in a file, do we mean (a) all diagnostics with a given code no matter which kind of fix is applied, (b) all diagnostics that can be fixed by a specific kind of fix no matter what their code is, or (c) all diagnostics with a given code that can be fixed by a specific kind of fix? I'd be very interested to hear your thoughts on this question.

> sdk/pkg/analysis_server/lib/src/lsp/handlers/handler_code_actions.dart
> Here, I think fix.change.id might also be the fix kind.

Yes, that's what I remembered as well. That's why you went through a while back and converted all of the ids to be dot-separated names. So if we do split some of the fix kinds we just need to be careful about the ids we given the new kinds so that they don't break users (or have a very compelling reason why we have to break users).

**Comment by DanTup** [comment]
> So, when we "fix all" in a file, do we mean (a) all diagnostics with a given code no matter which kind of fix is applied, (b) all diagnostics that can be fixed by a specific kind of fix no matter what their code is, or (c) all diagnostics with a given code that can be fixed by a specific kind of fix? I'd be very interested to hear your thoughts on this question.

I previously thought I wanted _b_ because it seems strange for the LSP "Apply all" to describe a fix, but actually fix all of an error code. However thinking more about it, I feel like it should be scoped to the specific error (_c_), because I think the intention is more likely "fix all instances of this error in this way".

That said, I think I've seen some errors where there are multiple diagnostic codes (perhaps to have more specific error messages) that represent the same logical kind of error, in which case it might be frustrating to seem like you're having to apply the same fix to the whole file multiple times.

> So if we do split some of the fix kinds we just need to be careful about the ids we given the new kinds so that they don't break users (or have a very compelling reason why we have to break users).

Yep, I think if we generally try to update them by adding dot-suffixes (which probably works well if they're being split into more-specific versions) then there should be no impact on users since the original filter would still match.

**Comment by bwilkerson** [comment]
> That said, I think I've seen some errors where there are multiple diagnostic codes (perhaps to have more specific error messages) that represent the same logical kind of error, in which case it might be frustrating to seem like you're having to apply the same fix to the whole file multiple times.

Yes, they exist for the purpose of allowing more specific messages. Diagnostic codes actually have two names at the moment: the "unique name", which is used to re-create the diagnostic from the serialized data that server stores, and the "name", which is used for documentation purposes. We might want to add a third value that specifies the "logical kind of error" in a way that makes the fix support work more like the way users expect it to.
